R/register_handlers.R

```diff
@@ -0,0 +1,98 @@
+#' Register handlers for logging messages, warnings and errors
+#'
+#' @param namespace (`character(1)`) the logger namespace
+#' @param package (`character(1)`) the package name
+#'
+#' @return `NULL` invisibly. Called for its side effects.
+#'
+#' @details This function registers global handlers for messages, warnings and errors.
+#' The handlers will investigate the call stack and if it contains a function
+#' from the package, the message, warning or error will be logged into the respective
+#' namespace.
+#' The handlers are registered only once per package and type. Consecutive calls will be ignored.
+#'
+#' @note Registering handlers is disallowed within `tryCatch` blocks which is used throughout
+#' base R and other packages. Because of this, the `register_handlers` function is wrapped in a
+#' `try` block to avoid errors.
+#'
+#' @seealso [globalCallingHandlers()]
+#'
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' register_logger(namespace = "teal.logger")
+#' register_handlers("teal.logger")
+#' # see the outcome
+#' globalCallingHandlers()
+#' }
+register_handlers <- function(namespace, package = namespace) {
+  try(
+    {
+      register_handler_type(namespace = namespace, package = package, type = "message")
+      register_handler_type(namespace = namespace, package = package, type = "warning")
+      register_handler_type(namespace = namespace, package = package, type = "error")
+    },
+    silent = TRUE
+  )
+
+  invisible(NULL)
+}
+
+#' @keywords internal
+register_handler_type <- function(
+    namespace,
+    package,
+    type = c("error", "warning", "message")) {
+  match.arg(type)
+  if (!(is.character(namespace) && length(namespace) == 1)) {
+    stop("namespace argument must be a scalar character.")
+  }
```

Comment:
```suggestion
  checkmate::assert_string(namespace)
```

Reply:
I would not use `scalar` in terms of R. There are no scalars. Only vectors of length 1. With checkmate this is a short test, that has a meaningful message that detects the type and the lenght.

Reply:
Unfortunately `checkmate` is not there yet. I didn't want to explode this PR by introducing this. That should be probably a separate issue and PR - also easier to review and track changes.

Re: "scalar". You are definitely right. This is just to match existing error messages. (below is current main but a different function)
https://github.com/insightsengineering/teal.logger/blob/497337faa03896da94c341a641490401540e5e70/R/register_logger.R#L54-L56.

I will try to improve it to be more correct

Reply:
Sorry, I thought `checkmate` is used across all `teal.*` packages. However in here we only have 2 functions in the package and we do not use `checkmate` for the verification. Ok, stay with a check without `checkmate.`

You can consider `stopifnot(` which takes named input to throw messages. It's an alternative for `if(!` + `stop`.


```r

type <- "b"

stopifnot(
  "message" = condition
)

condition <- is.numeric(type)

stopifnot(
  "message" = condition
)
> Error: message
```

Reply:
That is also true, but again - just wanted to "fit" into the existing conventions. See aforementioned function and the way how it validates arguments.
I do prefer checkmate and all of that but I think it's important to assure coherence in the package codebase and don't try to force individual preferences. At least - not that much :)

Reply:
sure, so ignore